django/db/models/query.py

```diff
@@ -1606,8 +1609,11 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
                 setattr(obj, to_attr, vals)
                 obj._prefetched_objects_cache[cache_name] = vals
             else:
-                # Cache in the QuerySet.all().
-                qs = getattr(obj, to_attr).all()
+                manager = getattr(obj, to_attr)
+                if leaf and lookup.queryset is not None:
+                    qs = manager._filter_queryset(lookup.queryset)
```

Comment:
I don't see fixing `model.relation.all().all()` as much of a backwards incompatibility, but the introduction of `_filter_queryset` is more troublesome. I haven't checked but I suspect projects that do fishy stuff with relation managers like taggit or mptt may hit an `AttributeError`. I can imagine 3 options:
1. Add `hasattr` to the if clause.
2. Leave it as it is and document the incompatibility, custom related managers are more or less private APIs and it's easy to fix.
3. Issue a deprecation warning when `_filter_queryset` is missing then switch to 2. after the deprecation.

I'm fine with both 2. and 3., 1. doesn't give a chance to projects to fix this inconsistency in their code.

Btw the relation manager of `GenericRelation` probably needs a similar treatment.